### PR TITLE
Add: mlflow for experiment tracking

### PR DESCRIPTION
**Pull request by JohanWork**
Adding MLFOW to Axolotl for experiment tracking, looked into how Weight and Bias has been setup and tried to follow the same pattern. Have tested the changes and everything looks good to me. 

Happy for any feedback or comments.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -694,6 +694,10 @@ wandb_name: # Set the name of your wandb run
 wandb_run_id: # Set the ID of your wandb run
 wandb_log_model: # "checkpoint" to log model to wandb Artifacts every `save_steps` or "end" to log only at the end of training
 
+# mlflow configuration if you're using it
+# Make sure your `MLFLOW_TRACKING_URI` is set.
+mlflow_experiment_name: # Your experiment name
```
Comment (winglian) [inline README.md:699]: It might be worth logging a warning when this is set but the uri env isnt.

**Review by winglian — APPROVED** [review]
Thanks for the PR!

File: README.md (hunk 1 above)
Comment (NanoCode012) [reply README.md:699]: Wouldn't it be better to make a config `mlflow_tracking_uri: `? The parsing code should be able to set that if needed.

Comment (JohanWork) [reply README.md:699]: I agree with what you propose @NanoCode012 . Have update the code.

**Comment by NanoCode012** [comment]
Is there any need to pass this config to the HF Trainer to enable log? I think some validation test would also be needed to make sure both aren't active at same time.

In addition, would you be able to provide an example run output + image on mlflow?

**Comment by winglian** [comment]
I disagree, isn't the tracking uri equivalent to a secret token? It can get inadvertently exposed if you share your YAML (say integration with wandb also)

**Comment by JohanWork** [comment]
> I disagree, isn't the tracking uri equivalent to a secret token? It can get inadvertently exposed if you share your YAML (say integration with wandb also)

To my understanding it is only an url actually. In Mlflow auth seams to be handled with username and password. But I might miss something. For reference https://mlflow.org/docs/latest/auth/index.html 

I have no strong opinion if it should be or not be in the config file.

**Comment by NanoCode012** [comment]
Oh, I wasn't aware it was a secret. In this case, I would not recommend it being in the yaml anymore due to reasons wing mentioned.

**Comment by winglian** [comment]
looks like the URI isn't considered a secret, but could be sensitive if someone starts a server and doesn't enable auth on it.
`MLFLOW_TRACKING_USERNAME` and `MLFLOW_TRACKING_PASSWORD` are the respective credentials

File: src/axolotl/core/trainer_builder.py [hunk 2]
```diff
@@ -738,7 +738,9 @@ def build(self, total_num_steps):
             False if self.cfg.ddp else None
         )
         training_arguments_kwargs["group_by_length"] = self.cfg.group_by_length
-        training_arguments_kwargs["report_to"] = "wandb" if self.cfg.use_wandb else None
+        training_arguments_kwargs["report_to"] =  (
+            "wandb" if self.cfg.use_wandb else 'mlflow' if self.cfg.use_mlflow else None
```
Comment (NanoCode012) [inline src/axolotl/core/trainer_builder.py]: Would prefer this be extracted out to make it more readable.

File: src/axolotl/core/trainer_builder.py [hunk 3]
```diff
@@ -738,7 +738,9 @@ def build(self, total_num_steps):
             False if self.cfg.ddp else None
         )
         training_arguments_kwargs["group_by_length"] = self.cfg.group_by_length
-        training_arguments_kwargs["report_to"] = "wandb" if self.cfg.use_wandb else None
+        training_arguments_kwargs["report_to"] = (
+            "wandb" if self.cfg.use_wandb else "mlflow" if self.cfg.use_mlflow else None
+        )
```
Comment (winglian) [inline src/axolotl/core/trainer_builder.py]:
```suggestion
        report_to = None
        if self.cfg.use_wandb:
            report_to = "wandb"
        if self.cfg.use_mlflow:
            report_to = "mlflow"
        training_arguments_kwargs["report_to"] = report_to
```
@NanoCode012 like this?

**Pull request merged**